# Introduction to an Encryption Model Using ASCII Manipulation
**Created by Carrillo and Francisco** 

This model leverages a combination of split and process technique (ensembling divide and conquer) and manipulation of ASCII values to enhance data security. By splitting the input data into smaller segments, this approach adds layers of complexity to the encryption process.

Central to this model is the manipulation of ASCII values, which allows for a dynamic transformation of the original data. By applying mathematical operations to the ASCII values of characters, we create a unique mapping that jumbles the plaintext, making it resistant to common cryptographic attacks. The purpose of this model is that even if an attacker gains access to a portion of the encrypted data, deciphering the entire message should remain quite a challenge.

The scope of this model accepts user input only within the range of characters with ASCII value of 32 to 126, with a maximum of 20 characters per input. The first 32 characters in the ASCII-table are unprintable control codes and are used to control peripherals such as printers. Codes 32-127 are common for all the different variations of the ASCII table, they are called printable characters, represent letters, digits, punctuation marks, and a few miscellaneous symbols. However, character 127 represents the command DEL. 

For this model, we'll be using dataframes for visualization and demonstration of the encryption formulation and process.

In [75]:
import numpy as np
import pandas as pd
import random as r
import math as m

In [76]:
df = pd.read_csv('message_dataset.csv')

For our example, we created a dataset with a list of messages and their expanded forms where each character has their corresponding indices.

In [77]:
df

Message  0  1  2  3
0    fish  f  i  s  h
1    book  b  o  o  k
2    star  s  t  a  r
3    moon  m  o  o  n
4    rock  r  o  c  k
5    dogs  d  o  g  s
6    love  l  o  v  e
7    tree  t  r  e  e
8    wind  w  i  n  d
9    play  p  l  a  y

In this part, we generated another 4 columns depicting the ASCII values of each characters in the message.

In [78]:
for row in range(10):  
    for col in range(4): 
        ascii_column = f'ASCII[{col}]'
        df.at[row, ascii_column] = ord(df.at[row, str(col)])

In [79]:
df

Message  0  1  2  3  ASCII[0]  ASCII[1]  ASCII[2]  ASCII[3]
0    fish  f  i  s  h     102.0     105.0     115.0     104.0
1    book  b  o  o  k      98.0     111.0     111.0     107.0
2    star  s  t  a  r     115.0     116.0      97.0     114.0
3    moon  m  o  o  n     109.0     111.0     111.0     110.0
4    rock  r  o  c  k     114.0     111.0      99.0     107.0
5    dogs  d  o  g  s     100.0     111.0     103.0     115.0
6    love  l  o  v  e     108.0     111.0     118.0     101.0
7    tree  t  r  e  e     116.0     114.0     101.0     101.0
8    wind  w  i  n  d     119.0     105.0     110.0     100.0
9    play  p  l  a  y     112.0     108.0      97.0     121.0

Now, we split the overall length of the message and apply different formula or calculation for each halves.

In [80]:
df['1st_half'] = df['0'] + df['1']
df['2nd_half'] = df['2'] + df['3']
df['ASCII[1st_half]'] = df['ASCII[0]'].astype(str) + ', ' + df['ASCII[1]'].astype(str)
df['ASCII[2nd_half]'] = df['ASCII[2]'].astype(str) + ', ' + df['ASCII[3]'].astype(str)

Now that we split the characters and their ascii values and grouped them into halves, we can now start our calculation. In our model, each half has different arithmetic formula or treatment for their shifts. We also used a number randomizer ranging from 1 to 32, that will be used in creating our shift values. This creates a total number of 32 possible encryptions for each input.



In [81]:
def randomize_number(min,max):
    return r.randint(min,max)

min_base = 1
max_base = 32
random_number = randomize_number(min_base,max_base)
random_number

28

Now that we've obtained a random number, we can now formulate our shift value. Below is the code for the shift_value:

In [82]:
# where n is the length of input; 
# and midpoint is the middle number after splitting the characters for each half of the string:

print("Random number: " , random_number)
def encrypt_half(message, is_first_half):
    n = len(message)
    shift_keys=[]
    shift_keys2=[]

    for i in range(n):
       
        shift_value = random_number + i 
        if shift_value % 2 == 0: 
            if is_first_half:
                final_shift = shift_value + (i + (i ** 2))
                shift_keys.append(final_shift)
            else:
                final_shift = shift_value + 2 * (i + (i ** 2))
                shift_keys2.append(final_shift)
        else: 
            if is_first_half:
                final_shift = shift_value + (i + (i ** 3))
                shift_keys.append(final_shift)
            else:
                final_shift = shift_value + 2 * (i + (i ** 3))
                shift_keys2.append(final_shift)

        print(f"[Encrypt] Char: {message[i]}, Final Shift: {final_shift}")
        column = 'Final Shift[first_half]' if is_first_half else 'Final Shift[2nd_half]'
        df.at[0, column] = str(shift_keys) if is_first_half else str(shift_keys2)
        
sample_message = df['Message'].iloc[0]
midpoint = len(sample_message) // 2
first_half_of_the_input = sample_message[:midpoint]
second_half_of_the_input = sample_message[midpoint:]

encrypt_half(first_half_of_the_input,is_first_half=True)
encrypt_half(second_half_of_the_input,is_first_half=False)

Random number:  28
[Encrypt] Char: f, Final Shift: 28
[Encrypt] Char: i, Final Shift: 31
[Encrypt] Char: s, Final Shift: 28
[Encrypt] Char: h, Final Shift: 33



### Shift Value Formula

The shift value contains the sum of random number generated and the current index of the character of a half.

### shift_value = random_number + i 

sample message: fish

random_number = 32

        for the first half of the message:

####        f 0 = 32 + 0 = 32

####        i 1 = 32 + 1 = 33 

        for the second half of the message

####        s 0 = 32 + 0 = 32

####        h 1 = 32 + 1 = 33 



### Final Shift Value Formula

The final shift value contains the sum of the shift value and the square or cube of the index added by itself then multiplied by one or two, depending on which half the character is in and if its obtained shift value is odd or even:

### if it is the first half, then:

            if shift_value is even, then: final_shift_value = shift_value + (i + (i^2))
    
            if shift_value is odd, then: final_shift_value = shift_value + (i + (i^3))

### if it is the second half, then:

            if shift_value is even, then: final_shift_value = shift_value + 2*(i + (i^2))

            if shift_value is odd, then: final_shift_value = shift_value + 2*(i + (i^3))


sample message: fish

### if it is the first half, then:

        if shift_value is even, then: 

            f [0] = 32 + (0 + 0^2) = 32
    
        if shift_value is odd, then:

            i [1] = 33 + (1 + 1^3) = 35

### if it is the second half, then:

        if shift_value is even, then: 

            s [0] = 32 + 2(0 + 0^2) = 32
    
        if shift_value is odd, then:
    
            h [1] = 33 + 2(1 + 1^3) = 37


#### Note:
Since the computation for the shift value is the sum of random_number and current index, 
the shift value will always be even, if:
 
        the character's current index == even and random_number == even
        the character's current index == odd and random_number == odd
 



In [83]:
df.drop(index=range(1,10))

Message  0  1  2  3  ASCII[0]  ASCII[1]  ASCII[2]  ASCII[3] 1st_half  \
0    fish  f  i  s  h     102.0     105.0     115.0     104.0       fi   

  2nd_half ASCII[1st_half] ASCII[2nd_half] Final Shift[first_half]  \
0       sh    102.0, 105.0    115.0, 104.0                [28, 31]   

  Final Shift[2nd_half]  
0              [28, 33]

After formulating each shift value for every characters in an input, we can now calculate for the final ASCII value for our encrypted keys and ciphered text. The calculation is as follows:

#### ascii_value = (ord(sample_message[i]) + final_shift) - round(m.log(final_shift))

where: 
- (ord(sample_message[i]) is a py func that gets the ASCII value of each character in the message based on the ASCII table;
- final_shift is the shift we calculated earlier based on certain conditions (is_first_half, is_even); and
- round(m.log(final_shift) from the math package, which obtains only the rounded whole number of the natural logarithm of the acquired final_shift value


### For encryption calculation:

example:

ascii_value = (ord(sample_message[i]) + final_shift) - round(m.log(final_shift)

sample message: fish

f = ascii_value = (102 + 32) - int(log(32)) = 134 - 3 = 131

#### while ascii_value > 126:
            ascii_value = 31 + (ascii_value - 126)

#### then:
            encrypted_key = 31 + (131 - 126) = 36            

for the character 'f' in the message 'fish', the calculated encrypted key of our model is 36.

### Reverse arithmetic of encryption (decryption calculation):

encrypted_key for 'f' =  36.

Input : f

Ciphered text : $

decrypted_value = (encrypted_keys[i] - final_shift) + round(m.log(final_shift))

decrypted_value = (36 - 32) + 3 = 7 

    while ascii_value < 32:
        ascii_value = 126 - (31 - decrypted_value)
        decrypted_value = 126 - (31 - 7) = 102

decrypted_key = 102

equivalent character for 102 = 'f'


In [84]:
# where n is the length of input; 
# and midpoint is the middle number after splitting the characters for each half of the string:

print("Random number: " , random_number)
def encrypt_half(message, is_first_half):
    n = len(message)
    ek_1=[]
    ek_2=[]
    encrypted_keys=[]

    for i in range(n):
       
        shift_value = random_number + i 
        if shift_value % 2 == 0: 
            if is_first_half:
                final_shift = shift_value + (i + (i ** 2))
                
            else:
                final_shift = shift_value + 2 * (i + (i ** 2))
                
        else: 
            if is_first_half:
                final_shift = shift_value + (i + (i ** 3))
                
            else:
                final_shift = shift_value + 2 * (i + (i ** 3))
                
        
        ascii_value = (ord(message[i]) + final_shift) - round(m.log(final_shift))
       
        while ascii_value > 126:
            ascii_value = 31 + (ascii_value - 126)
        
        ek_1.append(ascii_value) if is_first_half else ek_2.append(ascii_value)
        column = 'EncryptedKeys[1st]' if is_first_half else 'EncryptedKeys[2nd]'
        df.at[0, column] = str(ek_1) if is_first_half else str(ek_2)
        
    encrypted_keys = ek_1 + ek_2
    ciphered_text = ''.join(chr(key) for key in encrypted_keys)
    print("encrypted keys: ", encrypted_keys)
    print("Ciphered text: ", ciphered_text)    
    
sample_message = df['Message'].iloc[0]
midpoint = len(sample_message) // 2
first_half_of_the_input = sample_message[:midpoint]
second_half_of_the_input = sample_message[midpoint:]

encrypt_half(first_half_of_the_input,is_first_half=True)
encrypt_half(second_half_of_the_input,is_first_half=False)

Random number:  28
encrypted keys:  [32, 38]
Ciphered text:   &
encrypted keys:  [45, 39]
Ciphered text:  -'


In [85]:
df.drop(index=range(1,10))

Message  0  1  2  3  ASCII[0]  ASCII[1]  ASCII[2]  ASCII[3] 1st_half  \
0    fish  f  i  s  h     102.0     105.0     115.0     104.0       fi   

  2nd_half ASCII[1st_half] ASCII[2nd_half] Final Shift[first_half]  \
0       sh    102.0, 105.0    115.0, 104.0                [28, 31]   

  Final Shift[2nd_half] EncryptedKeys[1st] EncryptedKeys[2nd]  
0              [28, 33]           [32, 38]           [45, 39]

# Encryption Algorithm

This part demonstrates the code for the encryption model.

In [86]:
# import random as r
# import math as m

# def randomize_number(min,max):
#     return r.randint(min,max)

# min_base = 32
# max_base = 32
# random_number = randomize_number(min_base,max_base)

# def encrypt_half(message, is_first_half):
#     encrypted_keys = []
#     keys = []
    
#     n = len(message)
#     for i in range(n):
       
#         shift_temp = random_number + i 
#         if shift_temp % 2 == 0: 
#             if is_first_half:
#                 shift = shift_temp + (i + (i ** 2))
#             else:
#                 shift = shift_temp + 2 * (i + (i ** 2))
#         else: 
#             if is_first_half:
#                 shift = shift_temp + (i + (i ** 3))
#             else:
#                 shift = shift_temp + 2 * (i + (i ** 3))
    
#         ascii_value = (ord(message[i]) + shift) - round(m.log(shift))
       
#         while ascii_value > 126:
#             ascii_value = 31 + (ascii_value - 126)
        
#         #print(f"[Encrypt] Char: {message[i]}, Shift: {shift}, Encrypted ASCII: {ascii_value}")  # for tracking
   
#         encrypted_keys.append(ascii_value)
#         keys.append(ord(message[i]))

#     return keys, encrypted_keys

# def decrypt_half(encrypted_keys, is_first_half):
#     decrypted_message = ""
    
#     n = len(encrypted_keys)
#     for i in range(n):
#         shift_temp = random_number + i
#         if shift_temp % 2 == 0: 
#             if is_first_half:
#                 shift = shift_temp + (i + (i ** 2))
#             else:
#                 shift = shift_temp + 2 * (i + (i ** 2))
#         else:  
#             if is_first_half:
#                 shift = shift_temp + (i + (i ** 3))
#             else:
#                 shift = shift_temp + 2 * (i + (i ** 3))

#         ascii_value = (encrypted_keys[i] - shift) + round(m.log(shift))
        
#         while ascii_value < 32:
#             ascii_value = 126 - (31 - ascii_value)

#         decrypted_message += chr(ascii_value)

#     return decrypted_message

# def main():
#     while True:
#         action = input("Would you like to (e)ncrypt or (d)ecrypt a message? (q to quit): ").lower()
#         if action == 'e':

#             while True:
#                 message = input("Enter the message to encrypt: ")
#                 if len(message) <= 20:

#                     midpoint = len(message) // 2
                    
#                     first_half = message[:midpoint]
#                     second_half = message[midpoint:]

#                     keys1, encrypted_keys1 = encrypt_half(first_half, is_first_half=True)
#                     keys2, encrypted_keys2 = encrypt_half(second_half, is_first_half=False)

#                     final_keys = keys1 + keys2
#                     final_encrypted_keys = encrypted_keys1 + encrypted_keys2
#                     ciphered_text = ''.join(chr(key) for key in final_encrypted_keys)

#                     print("Original message:", message)
#                     # print("Keys (ASCII):", final_keys)
#                     # print("Encrypted keys:", final_encrypted_keys)
#                     print("Ciphered text:", ciphered_text)
#                     # print("Random number:", random_number)
#                     break
#                 else:
#                     print("ERROR: Limit reached. Shorten your message to 20 characters.")

#         elif action == 'd':

#             while True:
#                 encrypted_message = input("Enter the ciphered text to decrypt: ")

#                 if len(encrypted_message) <= 20:
#                     encrypted_keys = [ord(char) for char in encrypted_message]

#                     midpoint = len(encrypted_keys) // 2
                    
#                     first_half_encrypted = encrypted_keys[:midpoint]
#                     second_half_encrypted = encrypted_keys[midpoint:]

#                     decrypted_first_half = decrypt_half(first_half_encrypted, is_first_half=True)
#                     decrypted_second_half = decrypt_half(second_half_encrypted, is_first_half=False)

#                     decrypted_message = decrypted_first_half + decrypted_second_half
#                     print("Decrypted message:", decrypted_message)
#                     break
#                 else:
#                     print("ERROR: Limit reached. Shorten your message to 20 characters.")

#         elif action == 'q':
#             print("You have quit the program. Aww :((")
#             break
#         else:
#             print("ERROR: Invalid input, please choose 'e', 'd', or 'q'.")

# if __name__ == "__main__":
#     main()
